sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/PooledHashMap.java

```diff
@@ -0,0 +1,253 @@
+   * @param key the key whose associated value is to be returned
+   * @return The mapped value for {@code key} or null if there is no such mapping
+   */
+  @Override
+  @Nullable
+  @SuppressWarnings("unchecked")
+  public V get(Object key) {
+    int bucket = getBucket((K) key);
+    ArrayList<Entry<K, V>> entries = table[bucket];
+    if (entries != null) {
+      for (int i = 0; i < entries.size(); i++) {
+        Entry<K, V> entry = entries.get(i);
+        if (Objects.equals(entry.key, key)) {
+          return entry.value;
+        }
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Removes the mapping for the given {@code key}.
+   *
+   * @param key key whose mapping is to be removed from the map
+   * @return The value mapped to this key, if the mapping exists, or null otherwise
+   */
+  @Override
+  @Nullable
+  @SuppressWarnings("unchecked")
+  public V remove(Object key) {
+    int bucket = getBucket((K) key);
+    ArrayList<Entry<K, V>> entries = table[bucket];
+    if (entries != null) {
+      for (int i = 0; i < entries.size(); i++) {
+        Entry<K, V> entry = entries.get(i);
+        if (Objects.equals(entry.key, key)) {
+          V oldValue = entry.value;
+          entries.remove(i);
+          entryPool.returnObject(entry);
+          size--;
+          return oldValue;
+        }
+      }
+    }
+    return null;
+  }
+
+  @Override
+  public int size() {
+    return size;
+  }
+
+  @Override
+  public boolean isEmpty() {
+    return size == 0;
+  }
+
+  @Override
+  public boolean containsKey(Object key) {
+    return get(key) != null;
+  }
+
+  @Override
+  public boolean containsValue(Object value) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void clear() {
+    for (int i = 0; i < table.length; i++) {
+      ArrayList<Entry<K, V>> bucket = table[i];
+      if (bucket != null) {
+        for (int j = 0; j < bucket.size(); j++) {
+          Entry<K, V> entry = bucket.get(j);
+          entryPool.returnObject(entry);
+        }
+        bucket.clear();
+      }
+    }
+    size = 0;
+  }
+
+  @Override
+  public void forEach(BiConsumer<? super K, ? super V> action) {
+    for (int j = 0; j < table.length; j++) {
+      ArrayList<Entry<K, V>> bucket = table[j];
+      if (bucket != null) {
+        for (int i = 0; i < bucket.size(); i++) {
+          Entry<K, V> entry = bucket.get(i);
+          action.accept(entry.key, entry.value);
+        }
+      }
+    }
+  }
+
+  private int getBucket(K key) {
+    return Math.abs(key.hashCode() % table.length);
+  }
+
+  @Override
+  public Set<Map.Entry<K, V>> entrySet() {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public Collection<V> values() {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void putAll(Map<? extends K, ? extends V> m) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public Set<K> keySet() {
+    throw new UnsupportedOperationException();
```

Comment:
Might we worthwhile to add javadoc stating which common methods aren't supported, since I suspect other places in the codebase will want to use this.

Separately, I suppose these methods aren't implemented because: 1. They aren't used. 2. They would expose the state of map to callers in an unsafe way since the state of a given `Entry` can change.

Reply:
I was hesitant on adding a list of method to the javadoc of the class since I suspect as time goes by we may add implementation of one or two additional methods and I didn't that comment to go stale - WDYT?

Initially I implemented all method (well not me, chatGPT) but I decided to reduce cognitive load of readers like you in the PR and future readers of the code, to only keep methods that are actually used. This also reduces lots of test code.